# Why: The core-only engine library must not resolve neutral definitions from
# Dart-owned targets. Move view focus value definitions and shell
# initialization into core targets while keeping tonic's Dart logging handler
# in shell_dart.cc.
diff --git a/engine/src/flutter/lib/ui/BUILD.gn b/engine/src/flutter/lib/ui/BUILD.gn
index f64134fc20f..7d93d7b98bb 100644
--- a/engine/src/flutter/lib/ui/BUILD.gn
+++ b/engine/src/flutter/lib/ui/BUILD.gn
@@ -38,6 +38,7 @@ source_set("ui_core_types") {
     "window/viewport_metrics.cc",
     "window/viewport_metrics.h",
     "window/view_focus.h",
+    "window/view_focus.cc",
   ]
 
   public_configs = [ "//flutter:config" ]
@@ -318,7 +319,6 @@ source_set("ui") {
     "window/platform_message_response_dart.h",
     "window/platform_message_response_dart_port.cc",
     "window/platform_message_response_dart_port.h",
-    "window/view_focus.cc",
   ]
 
   public_configs = [ "//flutter:config" ]
diff --git a/engine/src/flutter/shell/common/shell.cc b/engine/src/flutter/shell/common/shell.cc
index 3f53cbe759f..1625bfeac59 100644
--- a/engine/src/flutter/shell/common/shell.cc
+++ b/engine/src/flutter/shell/common/shell.cc
@@ -3,6 +3,7 @@
 // found in the LICENSE file.
 
 #include <future>
+#include <mutex>
 #include "fml/task_runner.h"
 #include "impeller/core/runtime_types.h"
 #include "impeller/renderer/context.h"
@@ -19,6 +20,8 @@
 #include "flutter/common/graphics/persistent_cache.h"
 #include "flutter/fml/base32.h"
 #include "flutter/fml/file.h"
+#include "flutter/fml/icu_util.h"
+#include "flutter/fml/log_settings.h"
 #include "flutter/fml/logging.h"
 #include "flutter/fml/make_copyable.h"
 #include "flutter/fml/message_loop.h"
@@ -53,6 +56,63 @@ constexpr char kFontChange[] = "fontsChange";
 
 namespace {
 
+void RegisterCodecsWithSkia() {
+  // These are in the order they will be attempted to be decoded from.
+  // If we have data to back it up, we can order these by "frequency used in
+  // the wild" for a very small performance bump, but for now we mirror the
+  // order Skia had them in.
+  SkCodecs::Register(SkPngDecoder::Decoder());
+  SkCodecs::Register(SkJpegDecoder::Decoder());
+  SkCodecs::Register(SkWebpDecoder::Decoder());
+  SkCodecs::Register(SkGifDecoder::Decoder());
+  SkCodecs::Register(SkBmpDecoder::Decoder());
+  SkCodecs::Register(SkWbmpDecoder::Decoder());
+  SkCodecs::Register(SkIcoDecoder::Decoder());
+}
+
+}  // namespace
+
+void PerformInitializationTasks(Settings& settings) {
+  {
+    fml::LogSettings log_settings;
+    log_settings.min_log_level =
+        settings.verbose_logging ? fml::kLogInfo : fml::kLogError;
+    fml::SetLogSettings(log_settings);
+  }
+
+  static std::once_flag gShellSettingsInitialization = {};
+  std::call_once(gShellSettingsInitialization, [&settings] {
+    if (settings.trace_skia) {
+      InitSkiaEventTracer(settings.trace_skia, settings.trace_skia_allowlist);
+    }
+
+    if (!settings.trace_allowlist.empty()) {
+      fml::tracing::TraceSetAllowlist(settings.trace_allowlist);
+    }
+
+    if (!settings.skia_deterministic_rendering_on_cpu) {
+      SkGraphics::Init();
+    } else {
+      FML_DLOG(INFO) << "Skia deterministic rendering is enabled.";
+    }
+    RegisterCodecsWithSkia();
+
+    if (settings.icu_initialization_required) {
+      if (!settings.icu_data_path.empty()) {
+        fml::icu::InitializeICU(settings.icu_data_path);
+      } else if (settings.icu_mapper) {
+        fml::icu::InitializeICUFromMapping(settings.icu_mapper());
+      } else {
+        FML_DLOG(WARNING) << "Skipping ICU initialization in the shell.";
+      }
+    }
+  });
+
+#if !SLIMPELLER
+  PersistentCache::SetCacheSkSL(settings.cache_sksl);
+#endif  // !SLIMPELLER
+}
+
 std::unique_ptr<Engine> CreateEngine(
     Engine::Delegate& delegate,
     const std::shared_ptr<RuntimeBackendFactory>& runtime_backend_factory,
diff --git a/engine/src/flutter/shell/common/shell_dart.cc b/engine/src/flutter/shell/common/shell_dart.cc
index cb67a5cace9..95f21ebd939 100644
--- a/engine/src/flutter/shell/common/shell_dart.cc
+++ b/engine/src/flutter/shell/common/shell_dart.cc
@@ -8,10 +8,7 @@
 #include <mutex>
 #include <utility>
 
-#include "flutter/common/graphics/persistent_cache.h"
 #include "flutter/common/settings.h"
-#include "flutter/fml/icu_util.h"
-#include "flutter/fml/log_settings.h"
 #include "flutter/fml/logging.h"
 #include "flutter/fml/trace_event.h"
 #include "flutter/runtime/dart_vm.h"
@@ -19,81 +16,22 @@
 #include "flutter/shell/common/dart_runtime_backend_factory.h"
 #include "flutter/shell/common/dart_runtime/flutter_dart_runtime.h"
 #include "flutter/shell/common/shell.h"
-#include "flutter/shell/common/skia_event_tracer_impl.h"
-#include "third_party/skia/include/codec/SkBmpDecoder.h"
-#include "third_party/skia/include/codec/SkGifDecoder.h"
-#include "third_party/skia/include/codec/SkIcoDecoder.h"
-#include "third_party/skia/include/codec/SkJpegDecoder.h"
-#include "third_party/skia/include/codec/SkPngDecoder.h"
-#include "third_party/skia/include/codec/SkWbmpDecoder.h"
-#include "third_party/skia/include/codec/SkWebpDecoder.h"
-#include "third_party/skia/include/core/SkGraphics.h"
 #include "third_party/tonic/common/log.h"
 
 namespace flutter {
 
 namespace {
 
-void RegisterCodecsWithSkia() {
-  // These are in the order they will be attempted to be decoded from.
-  // If we have data to back it up, we can order these by "frequency used in
-  // the wild" for a very small performance bump, but for now we mirror the
-  // order Skia had them in.
-  SkCodecs::Register(SkPngDecoder::Decoder());
-  SkCodecs::Register(SkJpegDecoder::Decoder());
-  SkCodecs::Register(SkWebpDecoder::Decoder());
-  SkCodecs::Register(SkGifDecoder::Decoder());
-  SkCodecs::Register(SkBmpDecoder::Decoder());
-  SkCodecs::Register(SkWbmpDecoder::Decoder());
-  SkCodecs::Register(SkIcoDecoder::Decoder());
-}
-
-}  // namespace
-
-void PerformInitializationTasks(Settings& settings) {
-  {
-    fml::LogSettings log_settings;
-    log_settings.min_log_level =
-        settings.verbose_logging ? fml::kLogInfo : fml::kLogError;
-    fml::SetLogSettings(log_settings);
-  }
-
-  static std::once_flag gShellSettingsInitialization = {};
-  std::call_once(gShellSettingsInitialization, [&settings] {
+void InstallTonicLogHandler() {
+  static std::once_flag tonic_log_handler_initialization = {};
+  std::call_once(tonic_log_handler_initialization, [] {
     tonic::SetLogHandler(
         [](const char* message) { FML_LOG(ERROR) << message; });
-
-    if (settings.trace_skia) {
-      InitSkiaEventTracer(settings.trace_skia, settings.trace_skia_allowlist);
-    }
-
-    if (!settings.trace_allowlist.empty()) {
-      fml::tracing::TraceSetAllowlist(settings.trace_allowlist);
-    }
-
-    if (!settings.skia_deterministic_rendering_on_cpu) {
-      SkGraphics::Init();
-    } else {
-      FML_DLOG(INFO) << "Skia deterministic rendering is enabled.";
-    }
-    RegisterCodecsWithSkia();
-
-    if (settings.icu_initialization_required) {
-      if (!settings.icu_data_path.empty()) {
-        fml::icu::InitializeICU(settings.icu_data_path);
-      } else if (settings.icu_mapper) {
-        fml::icu::InitializeICUFromMapping(settings.icu_mapper());
-      } else {
-        FML_DLOG(WARNING) << "Skipping ICU initialization in the shell.";
-      }
-    }
   });
-
-#if !SLIMPELLER
-  PersistentCache::SetCacheSkSL(settings.cache_sksl);
-#endif  // !SLIMPELLER
 }
 
+}  // namespace
+
 std::pair<DartVMRef, fml::RefPtr<const DartSnapshot>>
 InferDartVmInitDataFromSettings(Settings& settings) {
   // Always use the `vm_snapshot` and `isolate_snapshot` provided by the
@@ -131,6 +69,7 @@ std::unique_ptr<Shell> Shell::Create(
     bool is_gpu_disabled) {
   // This must come first as it initializes tracing.
   PerformInitializationTasks(settings);
+  InstallTonicLogHandler();
 
   TRACE_EVENT0("flutter", "Shell::Create");
 
